File: android/app/src/main/java/com/vietbridge/ai/ui/home/ChatViewModel.kt
package com.vietbridge.ai.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vietbridge.ai.data.model.*
import com.vietbridge.ai.data.remote.ApiClient
import com.vietbridge.ai.data.remote.SSEEvent
import com.vietbridge.ai.data.remote.asSSEFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ChatState(
    val messages: List<ChatMessage> = emptyList(),
    val isStreaming: Boolean = false,
    val streamingContent: String = "",
    val error: String? = null,
    val selectedTask: String? = null,
    val selectedScene: String = "general",
    val langDir: String = "zh2vi",
    val tone: Int = 50,
    val conversationId: String? = null,
)

class ChatViewModel : ViewModel() {
    private val _state = MutableStateFlow(ChatState())
    val state: StateFlow<ChatState> = _state.asStateFlow()

    fun send(inputText: String) {
        val text = inputText.trim()
        if (text.isEmpty() || _state.value.isStreaming) return

        val userMsg = ChatMessage.user(text)
        _state.update { it.copy(
            messages = it.messages + userMsg,
            isStreaming = true,
            streamingContent = "",
            error = null,
        ) }

        val s = _state.value
        val history = s.messages.takeLast(10).map { HistoryMessage(it.role.name.lowercase(), it.content) }

        viewModelScope.launch {
            try {
                val request = ChatRequest(
                    input = text, task = s.selectedTask, scene = s.selectedScene,
                    langDir = s.langDir, tone = s.tone, stream = true,
                    conversationId = s.conversationId, conversationHistory = history,
                )
                val response = ApiClient.api.chatStream(request)
                val body = response.body() ?: throw Exception("空响应")

                var fullContent = ""
                var doneEvent: SSEDoneEvent? = null

                body.asSSEFlow().collect { event ->
                    when (event) {
                        is SSEEvent.Delta -> {
                            fullContent += event.content
                            _state.update { it.copy(streamingContent = fullContent) }
                        }
                        is SSEEvent.Done -> doneEvent = event.event
                        is SSEEvent.Error -> _state.update { it.copy(error = event.message) }
                    }
                }

                val assistantMsg = ChatMessage.assistant(
                    content = fullContent,
                    type = doneEvent?.messageType,
                    data = doneEvent?.data,
                    warnings = doneEvent?.proactiveWarnings,
                )
                _state.update { it.copy(
                    messages = it.messages + assistantMsg,
                    conversationId = doneEvent?.conversationId ?: it.conversationId,
                ) }
            } catch (e: Exception) {
                _state.update { it.copy(error = e.message ?: "发送失败") }
            }
            _state.update { it.copy(isStreaming = false, streamingContent = "") }
        }
    }

    fun newConversation() {
        _state.update { ChatState(selectedTask = it.selectedTask, selectedScene = it.selectedScene, langDir = it.langDir, tone = it.tone) }
    }

    fun setTask(task: String?) = _state.update { it.copy(selectedTask = task) }
    fun setScene(scene: String) = _state.update { it.copy(selectedScene = scene) }
    fun setLangDir(dir: String) = _state.update { it.copy(langDir = dir) }
    fun setTone(tone: Int) = _state.update { it.copy(tone = tone) }
}

File: ios/VietBridge/ViewModels/ChatViewModel.swift
import Foundation

@Observable
@MainActor
final class ChatViewModel {
    var messages: [ChatMessage] = []
    var loading = false
    var error: String?
    var task = "translate"
    var scene = "general"
    var tone = 50
    var langDir = "zh2vi"
    var conversationId: String?
    var detectedIntent: IntentInfo?

    func send(_ text: String) {
        guard !text.isEmpty, !loading else { return }
        messages.append(ChatMessage(type: .user, text: text))
        loading = true
        error = nil

        let streamMsg = ChatMessage(type: .streaming, text: "")
        messages.append(streamMsg)
        let streamIndex = messages.count - 1

        // Build conversation history from existing messages
        let history = messages.dropLast(2).map { m -> ConversationEntry in
            ConversationEntry(
                role: m.type == .user ? "user" : "assistant",
                content: m.text
            )
        }

        Task {
            do {
                var req = ChatRequest(
                    input: text, task: task, scene: scene, tone: tone,
                    stream: true, langDir: langDir
                )
                if let cid = conversationId {
                    req.conversationId = cid
                }
                if !history.isEmpty {
                    req.conversationHistory = Array(history)
                }

                let bytes = try await APIClient.shared.streamSSE("/api/chat", body: req)
                var accumulated = ""
                var finalData: ChatResponseData?
                var finalType: String?
                var warnings: [ProactiveWarning]?
                var kbHits: [KBHit]?

                for try await line in bytes.lines {
                    guard line.hasPrefix("data: ") else { continue }
                    let json = String(line.dropFirst(6))
                    guard let data = json.data(using: .utf8) else { continue }
                    let decoder = JSONDecoder()
                    guard let event = try? decoder.decode(SSEEvent.self, from: data) else { continue }

                    switch event.type {
                    case "delta":
                        accumulated += event.content ?? ""
                        messages[streamIndex].text = accumulated
                    case "done":
                        finalData = event.data
                        finalType = event.messageType
                        conversationId = event.conversationId ?? conversationId
                        warnings = event.proactiveWarnings
                        kbHits = event.kbHits
                        if let intent = event.intent,
                           let intentTask = intent.task,
                           let confidence = intent.confidence,
                           confidence >= 0.6,
                           intentTask != self.task {
                            self.detectedIntent = intent
                        } else {
                            self.detectedIntent = nil
                        }
                    case "error":
                        error = event.error ?? "请求失败"
                    default: break
                    }
                }

                messages[streamIndex] = ChatMessage(
                    type: .assistant,
                    text: accumulated,
                    responseType: finalType,
                    data: finalData,
                    proactiveWarnings: warnings,
                    kbHits: kbHits
                )
            } catch {
                self.error = error.localizedDescription
                if streamIndex < messages.count {
                    messages.remove(at: streamIndex)
                }
            }
            loading = false
        }
    }

    func clearMessages() {
        messages.removeAll()
        conversationId = nil
    }
}
